plugins/samples/add_device_type/plugin.go

```diff
@@ -0,0 +1,168 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// [START serviceextensions_plugin_device_type]
+package main
+
+import (
+	"strings"
+
+	"github.com/proxy-wasm/proxy-wasm-go-sdk/proxywasm"
+	"github.com/proxy-wasm/proxy-wasm-go-sdk/proxywasm/types"
+)
+
+const (
+	userAgentHeader  = "user-agent"    // The HTTP header containing the user agent string
+	deviceTypeHeader = "x-device-type" // The new HTTP header that will store the detected device type
+	defaultUserAgent = "unknown"       // Default value for user agent if not provided
+)
+
+const (
+	deviceTypeDesktop = "desktop"
+	deviceTypeTablet  = "tablet"
+	deviceTypePhone   = "phone"
+	deviceTypeBot     = "bot"
+	deviceTypeOther   = "other"
+)
+
+func main() {}
+
+type vmContext struct{ types.DefaultVMContext }
+type pluginContext struct{ types.DefaultPluginContext }
+type httpContext struct{ types.DefaultHttpContext }
+
+func init() { proxywasm.SetVMContext(&vmContext{}) }
+
+// NewPluginContext creates a new plugin context and returns it
+func (*vmContext) NewPluginContext(uint32) types.PluginContext {
+	return &pluginContext{}
+}
+
+// NewHttpContext creates a new HTTP context for handling HTTP requests
+func (*pluginContext) NewHttpContext(uint32) types.HttpContext {
+	return &httpContext{}
+}
+
+// OnHttpRequestHeaders is called when the HTTP request headers are received
+// It detects the device type based on the user agent and adds it as a new header
+func (ctx *httpContext) OnHttpRequestHeaders(int, bool) types.Action {
+	defer recoverPanic()
+
+	userAgent, err := proxywasm.GetHttpRequestHeader(userAgentHeader)
+	if err != nil {
+		handleUserAgent(err)
+		userAgent = defaultUserAgent
+	}
+
+	deviceType := detectDeviceType(strings.ToLower(userAgent))
+	// Set the new device type header in the request
+	if err := proxywasm.ReplaceHttpRequestHeader(deviceTypeHeader, deviceType); err != nil {
+		proxywasm.LogErrorf("failed to set device type header: %v", err)
+	}
+
+	return types.ActionContinue
+}
+
+// detectDeviceType determines the type of device based on the user agent string
+func detectDeviceType(ua string) string {
+	switch {
+	case isBot(ua):
+		return deviceTypeBot
+	case isTablet(ua):
+		return deviceTypeTablet
+	case isMobile(ua):
+		return deviceTypePhone
+	case isDesktop(ua):
+		return deviceTypeDesktop
+	default:
+		return deviceTypeOther
+	}
+}
+
+// isBot checks if the user agent indicates a bot or crawler
+func isBot(ua string) bool {
+	return containsAny(ua, []string{
```

Comment:
Suggest storing these lists in package-level variables so they don't need to be reallocated with each request